// falcon-ui/src/index.js
/* 基础组件 */
import Toast from './packages/toast/index.vue';
import FlRow from './packages/layout/row.vue';
import FlCol from './packages/layout/col.vue';
import FlIcon from './packages/icon/index.vue';
import FlButton from './packages/button/index.vue';

/* 表单组件 */
import FlCheckbox from './packages/checkbox/index.vue';
import FlSwitch from './packages/switch/index.vue';
import FlRadio from './packages/radio/index.vue';
import FlSeekbar from './packages/seekbar/index.vue';
import FlProgress from './packages/progress/index.vue';

/* 弹窗类组件 */
import FlModalInternal from './packages/modal-internal/index.vue';
import FlDialog from './packages/dialog/index.vue';
import FlPopup from './packages/popup/index.vue';

/* 二维码/一维码 */
import FlBarcode from './packages/bar-code/index.vue';
import FlQrcode from './packages/qr-code/index.vue';

/* 富文本组件 */
import FlRichText from './packages/rich-text/index.js';

/* lottie动画 */
import FlLottie from './packages/lottie/index.vue';

import FlSlider from './packages/slider/index.vue';
import FlImage from './packages/image/index.vue';

export {
  Toast,
  FlRow,
  FlCol,
  FlIcon,
  FlButton,

  FlCheckbox,
  FlSwitch,
  FlRadio,
  FlSeekbar,
  FlProgress,

  FlModalInternal,
  FlDialog,
  FlPopup,

  FlBarcode,
  FlQrcode,

  FlRichText,

  FlLottie,

  FlSlider,
  FlImage,
};
